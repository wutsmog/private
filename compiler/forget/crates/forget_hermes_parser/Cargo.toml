[package]
name = "forget_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
forget_diagnostics = { workspace = true }
forget_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
forget_estree_codegen = { workspace = true }