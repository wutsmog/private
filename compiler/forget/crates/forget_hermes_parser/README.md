# forget_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `forget_estree` AST.