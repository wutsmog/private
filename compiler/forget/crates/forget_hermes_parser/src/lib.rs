mod generated;
mod generated_extension;

use forget_diagnostics::Diagnostic;
use forget_estree::Program;
use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new();
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .into_iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
