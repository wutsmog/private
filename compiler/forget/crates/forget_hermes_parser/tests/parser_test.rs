use std::env;

use forget_hermes_parser::parse;
use insta::{assert_snapshot, glob};
use serde_json;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let ast = parse(&input, path.to_str().unwrap()).unwrap();
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
