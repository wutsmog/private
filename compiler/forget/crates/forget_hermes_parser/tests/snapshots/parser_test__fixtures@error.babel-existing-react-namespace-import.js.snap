---
source: crates/forget_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/forget_hermes_parser/tests/fixtures/error.babel-existing-react-namespace-import.js
---
Input:
import * as React from "react";

function Component(props) {
  const [x] = React.useState(0);
  const expensiveNumber = React.useMemo(() => calculateExpensiveNumber(x), [x]);

  return <div>{expensiveNumber}</div>;
}

function Component2(props) {
  const [x] = React.useState(0);
  const expensiveNumber = React.useMemo(() => calculateExpensiveNumber(x), [x]);

  return <div>{expensiveNumber}</div>;
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "ImportDeclaration",
      "specifiers": [
        {
          "type": "ImportNamespaceSpecifier",
          "local": {
            "type": "Identifier",
            "name": "React",
            "typeAnnotation": null,
            "loc": null,
            "range": {
              "start": 12,
              "end": 17
            }
          },
          "loc": null,
          "range": {
            "start": 7,
            "end": 17
          }
        }
      ],
      "source": {
        "type": "StringLiteral",
        "value": "react",
        "loc": null,
        "range": {
          "start": 23,
          "end": 30
        }
      },
      "loc": null,
      "range": {
        "start": 0,
        "end": 31
      }
    },
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "Component",
        "typeAnnotation": null,
        "loc": null,
        "range": {
          "start": 42,
          "end": 51
        }
      },
      "params": [
        {
          "type": "Identifier",
          "name": "props",
          "typeAnnotation": null,
          "loc": null,
          "range": {
            "start": 52,
            "end": 57
          }
        }
      ],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "ArrayPattern",
                  "elements": [
                    {
                      "type": "Identifier",
                      "name": "x",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 70,
                        "end": 71
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 69,
                    "end": 72
                  }
                },
                "init": {
                  "type": "CallExpression",
                  "callee": {
                    "type": "MemberExpression",
                    "object": {
                      "type": "Identifier",
                      "name": "React",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 75,
                        "end": 80
                      }
                    },
                    "property": {
                      "type": "Identifier",
                      "name": "useState",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 81,
                        "end": 89
                      }
                    },
                    "computed": false,
                    "loc": null,
                    "range": {
                      "start": 75,
                      "end": 89
                    }
                  },
                  "arguments": [
                    {
                      "type": "NumericLiteral",
                      "value": 0.0,
                      "loc": null,
                      "range": {
                        "start": 90,
                        "end": 91
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 75,
                    "end": 92
                  }
                },
                "loc": null,
                "range": {
                  "start": 69,
                  "end": 92
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 63,
              "end": 93
            }
          },
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "expensiveNumber",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 102,
                    "end": 117
                  }
                },
                "init": {
                  "type": "CallExpression",
                  "callee": {
                    "type": "MemberExpression",
                    "object": {
                      "type": "Identifier",
                      "name": "React",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 120,
                        "end": 125
                      }
                    },
                    "property": {
                      "type": "Identifier",
                      "name": "useMemo",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 126,
                        "end": 133
                      }
                    },
                    "computed": false,
                    "loc": null,
                    "range": {
                      "start": 120,
                      "end": 133
                    }
                  },
                  "arguments": [
                    {
                      "type": "ArrowFunctionExpression",
                      "id": null,
                      "params": [],
                      "body": {
                        "type": "CallExpression",
                        "callee": {
                          "type": "Identifier",
                          "name": "calculateExpensiveNumber",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 140,
                            "end": 164
                          }
                        },
                        "arguments": [
                          {
                            "type": "Identifier",
                            "name": "x",
                            "typeAnnotation": null,
                            "loc": null,
                            "range": {
                              "start": 165,
                              "end": 166
                            }
                          }
                        ],
                        "loc": null,
                        "range": {
                          "start": 140,
                          "end": 167
                        }
                      },
                      "generator": true,
                      "async": false,
                      "loc": null,
                      "range": {
                        "start": 134,
                        "end": 167
                      },
                      "expression": true,
                      "loc": null,
                      "range": {
                        "start": 134,
                        "end": 167
                      }
                    },
                    {
                      "type": "ArrayExpression",
                      "elements": [
                        {
                          "type": "Identifier",
                          "name": "x",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 170,
                            "end": 171
                          }
                        }
                      ],
                      "loc": null,
                      "range": {
                        "start": 169,
                        "end": 172
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 120,
                    "end": 173
                  }
                },
                "loc": null,
                "range": {
                  "start": 102,
                  "end": 173
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 96,
              "end": 174
            }
          },
          {
            "type": "ReturnStatement",
            "argument": {
              "type": "JSXElement",
              "openingElement": {
                "type": "JSXOpeningElement",
                "name": {
                  "type": "JSXIdentifier",
                  "name": "div",
                  "loc": null,
                  "range": {
                    "start": 186,
                    "end": 189
                  }
                },
                "attributes": [],
                "selfClosing": false,
                "loc": null,
                "range": {
                  "start": 185,
                  "end": 190
                }
              },
              "children": [
                {
                  "type": "JSXExpressionContainer",
                  "expression": {
                    "type": "Identifier",
                    "name": "expensiveNumber",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 191,
                      "end": 206
                    }
                  },
                  "loc": null,
                  "range": {
                    "start": 190,
                    "end": 207
                  }
                }
              ],
              "closingElement": {
                "type": "JSXClosingElement",
                "name": {
                  "type": "JSXIdentifier",
                  "name": "div",
                  "loc": null,
                  "range": {
                    "start": 209,
                    "end": 212
                  }
                },
                "loc": null,
                "range": {
                  "start": 207,
                  "end": 213
                }
              },
              "loc": null,
              "range": {
                "start": 185,
                "end": 213
              }
            },
            "loc": null,
            "range": {
              "start": 178,
              "end": 214
            }
          }
        ],
        "loc": null,
        "range": {
          "start": 59,
          "end": 216
        }
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": {
        "start": 33,
        "end": 216
      },
      "loc": null,
      "range": {
        "start": 33,
        "end": 216
      }
    },
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "Component2",
        "typeAnnotation": null,
        "loc": null,
        "range": {
          "start": 227,
          "end": 237
        }
      },
      "params": [
        {
          "type": "Identifier",
          "name": "props",
          "typeAnnotation": null,
          "loc": null,
          "range": {
            "start": 238,
            "end": 243
          }
        }
      ],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "ArrayPattern",
                  "elements": [
                    {
                      "type": "Identifier",
                      "name": "x",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 256,
                        "end": 257
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 255,
                    "end": 258
                  }
                },
                "init": {
                  "type": "CallExpression",
                  "callee": {
                    "type": "MemberExpression",
                    "object": {
                      "type": "Identifier",
                      "name": "React",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 261,
                        "end": 266
                      }
                    },
                    "property": {
                      "type": "Identifier",
                      "name": "useState",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 267,
                        "end": 275
                      }
                    },
                    "computed": false,
                    "loc": null,
                    "range": {
                      "start": 261,
                      "end": 275
                    }
                  },
                  "arguments": [
                    {
                      "type": "NumericLiteral",
                      "value": 0.0,
                      "loc": null,
                      "range": {
                        "start": 276,
                        "end": 277
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 261,
                    "end": 278
                  }
                },
                "loc": null,
                "range": {
                  "start": 255,
                  "end": 278
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 249,
              "end": 279
            }
          },
          {
            "type": "VariableDeclaration",
            "kind": "const",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "expensiveNumber",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 288,
                    "end": 303
                  }
                },
                "init": {
                  "type": "CallExpression",
                  "callee": {
                    "type": "MemberExpression",
                    "object": {
                      "type": "Identifier",
                      "name": "React",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 306,
                        "end": 311
                      }
                    },
                    "property": {
                      "type": "Identifier",
                      "name": "useMemo",
                      "typeAnnotation": null,
                      "loc": null,
                      "range": {
                        "start": 312,
                        "end": 319
                      }
                    },
                    "computed": false,
                    "loc": null,
                    "range": {
                      "start": 306,
                      "end": 319
                    }
                  },
                  "arguments": [
                    {
                      "type": "ArrowFunctionExpression",
                      "id": null,
                      "params": [],
                      "body": {
                        "type": "CallExpression",
                        "callee": {
                          "type": "Identifier",
                          "name": "calculateExpensiveNumber",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 326,
                            "end": 350
                          }
                        },
                        "arguments": [
                          {
                            "type": "Identifier",
                            "name": "x",
                            "typeAnnotation": null,
                            "loc": null,
                            "range": {
                              "start": 351,
                              "end": 352
                            }
                          }
                        ],
                        "loc": null,
                        "range": {
                          "start": 326,
                          "end": 353
                        }
                      },
                      "generator": true,
                      "async": false,
                      "loc": null,
                      "range": {
                        "start": 320,
                        "end": 353
                      },
                      "expression": true,
                      "loc": null,
                      "range": {
                        "start": 320,
                        "end": 353
                      }
                    },
                    {
                      "type": "ArrayExpression",
                      "elements": [
                        {
                          "type": "Identifier",
                          "name": "x",
                          "typeAnnotation": null,
                          "loc": null,
                          "range": {
                            "start": 356,
                            "end": 357
                          }
                        }
                      ],
                      "loc": null,
                      "range": {
                        "start": 355,
                        "end": 358
                      }
                    }
                  ],
                  "loc": null,
                  "range": {
                    "start": 306,
                    "end": 359
                  }
                },
                "loc": null,
                "range": {
                  "start": 288,
                  "end": 359
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 282,
              "end": 360
            }
          },
          {
            "type": "ReturnStatement",
            "argument": {
              "type": "JSXElement",
              "openingElement": {
                "type": "JSXOpeningElement",
                "name": {
                  "type": "JSXIdentifier",
                  "name": "div",
                  "loc": null,
                  "range": {
                    "start": 372,
                    "end": 375
                  }
                },
                "attributes": [],
                "selfClosing": false,
                "loc": null,
                "range": {
                  "start": 371,
                  "end": 376
                }
              },
              "children": [
                {
                  "type": "JSXExpressionContainer",
                  "expression": {
                    "type": "Identifier",
                    "name": "expensiveNumber",
                    "typeAnnotation": null,
                    "loc": null,
                    "range": {
                      "start": 377,
                      "end": 392
                    }
                  },
                  "loc": null,
                  "range": {
                    "start": 376,
                    "end": 393
                  }
                }
              ],
              "closingElement": {
                "type": "JSXClosingElement",
                "name": {
                  "type": "JSXIdentifier",
                  "name": "div",
                  "loc": null,
                  "range": {
                    "start": 395,
                    "end": 398
                  }
                },
                "loc": null,
                "range": {
                  "start": 393,
                  "end": 399
                }
              },
              "loc": null,
              "range": {
                "start": 371,
                "end": 399
              }
            },
            "loc": null,
            "range": {
              "start": 364,
              "end": 400
            }
          }
        ],
        "loc": null,
        "range": {
          "start": 245,
          "end": 402
        }
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": {
        "start": 218,
        "end": 402
      },
      "loc": null,
      "range": {
        "start": 218,
        "end": 402
      }
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": {
    "start": 0,
    "end": 402
  }
}
